### PR TITLE
feat(fetch): fetch with request parameter

### PR DESCRIPTION
**Pull request by yury-s**
#5999
#1774

### CODE REVIEW

File: src/client/browserContext.ts [hunk 1]
```diff
@@ -216,14 +216,23 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
     });
   }
 
-  async _fetch(url: string, options: FetchOptions = {}): Promise<network.FetchResponse> {
+  async _fetch(request: network.Request, options?: FetchOptions & { timeout?: number }): Promise<network.FetchResponse>;
+  async _fetch(url: string, options?: FetchOptions): Promise<network.FetchResponse>;
+  async _fetch(urlOrRequest: string|network.Request, options: FetchOptions = {}): Promise<network.FetchResponse> {
     return this._wrapApiCall(async (channel: channels.BrowserContextChannel) => {
-      const postDataBuffer = isString(options.postData) ? Buffer.from(options.postData, 'utf8') : options.postData;
+      const request: network.Request | undefined = (typeof urlOrRequest === 'object') ? urlOrRequest as network.Request : undefined;
```
Comment (dgozman) [inline src/client/browserContext.ts:223]: Can we do `instanceof network.Request` here? This way we'll get a nice error message if they we something strange. Perhaps add a test?

File: src/client/browserContext.ts [hunk 2]
```diff
@@ -216,14 +216,23 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
...
+      const url = request ? request.url() : urlOrRequest as string;
+      const method = request?.method() || options.method;
+      // Cannot call allHeaders() here as the request may be paused inside route handler.
```
Comment (dgozman) [inline src/client/browserContext.ts:226]: Ouch!!!

File: src/client/browserContext.ts [hunk 3]
```diff
@@ -216,14 +216,23 @@ export class BrowserContext extends ChannelOwner<channels.BrowserContextChannel,
...
+      const headersObj = request?.headers() || options.headers;
+      const headers = headersObj ? headersObjectToArray(headersObj) : undefined;
+      const postDataBuffer = request?.postDataBuffer() || (isString(options.postData) ? Buffer.from(options.postData, 'utf8') : options.postData);
```
Comment (dgozman) [inline src/client/browserContext.ts:229]: `request?.postDataBuffer()` may return null, in which case you use options?

**Review by dgozman — APPROVED** [review]

File: src/client/browserContext.ts (hunk 2 above)
Comment (yury-s) [reply src/client/browserContext.ts:226]: Yeah, it's a bummer.

File: src/client/browserContext.ts (hunk 1 above)
Comment (yury-s) [reply src/client/browserContext.ts:223]: Done.

File: src/client/browserContext.ts (hunk 3 above)
Comment (yury-s) [reply src/client/browserContext.ts:229]: Fixed.

**Pull request merged**